### ISSUE TITLE
hyperparameters tunning using rayTune

### ISSUE THREAD

**Issue by akashAD98** [ultralytics/ultralytics#3310]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I have 2 q.
1) how can i set metrix as val loss,because  my model is always giving me higher MAP, so I want to find the best parameters based on val loss,& higher accuracy
```
print(result.get_best_result(metric = 'val/seg_loss', mode = 'min'))
```
my code
```
from ultralytics import YOLO
from ray import tune

model = YOLO("yolov8n.pt")
result = model.tune(
    data="data_weapons4_mask.yaml",
    space={"lr0": tune.uniform(1e-5, 1e-1),
    "box": tune.uniform(0.02, 0.2),
    "cls": tune.uniform(0.2, 4.0)},
    max_samples=4, # num_samples=20,
    gpu_per_trial=1,
    train_args={"epochs": 20}

)


print(result.get_best_result(metric = 'val/seg_loss', mode = 'min'))
```
im getting error,such as this is not matrix avaialble

2. i saw the raytune documentation & it has differrent methods for serching & earlystopping, when i was trying to do its not working,
is there any way to implement this features or yolov8 dont support this?

```


from ultralytics import YOLO
from ray import tune
from ray.tune.search.bayesopt import BayesOptSearch
from ray.tune.schedulers import HyperBandScheduler

model = YOLO("yolov8n.pt")

#search algo
algo = BayesOptSearch(random_search_steps=4)

# earlystopping algo
# Create HyperBand scheduler and minimize the score
hyperband = HyperBandScheduler(metric="metrics/mAP50(B)", mode="max")

result = model.tune(
    data="coco128.yaml",

    space={"lr0": tune.uniform(1e-5, 1e-1),
    "box": tune.uniform(0.02, 0.2),
    "cls": tune.uniform(0.2, 4.0)},
    max_samples=2,
    train_args={"epochs": 10}

)


```

**Comment by pderrenger** [comment]
@akashAD98 hi there! 🌟

It seems like you're facing some issues with hyperparameter tuning using Ray Tune for YOLOv8. Let's address your questions step by step.

1) To set the metric as validation loss for hyperparameter tuning, you can utilize the functionality provided by Ultralytics YOLO. Specifically, you would define the "val/seg_loss" as the metric to optimize during tuning. However, it's important to note that while the `model.tune()` method automatically optimizes based on validation loss, the specific method `result.get_best_result(metric='val/seg_loss', mode='min')` might not be available directly as an attribute.

2) As for the second question, YOLOv8 does support numerous hyperparameter optimization algorithms and early stopping techniques. The `tune()` method allows you to specify the search algorithm and scheduler as you've done in your code. However, it's crucial to make sure that the metrics/mAP50(B) used in the HyperBandScheduler are consistent with the metric names used in the YOLOv8 training, validation, and evaluation processes.

Ensure that the metric names align with those defined and logged in the training setup to enable meaningful hyperparameter optimization and early stopping strategies.

If you encounter specific errors or unexpected behaviors, feel free to provide additional details, and we can further assist in troubleshooting. Happy tuning! 😊
